dask_expr/io/io.py

```diff
@@ -199,12 +205,19 @@ def _simplify_up(self, parent):
     def _divisions(self):
         return self._divisions_and_locations[0]
 
+    @functools.cached_property
+    def npartitions(self):
+        if self._filtered:
+            return super().npartitions
+        return len(self._divisions_and_locations[0]) - 1
+
```

Comment:
Maybe we can avoid adding this (which I find slightly confusion), and instead fix the source of the actual problem. As far as I understand it, the underlying problem is that [PartitionsFiltered.npartitions is incorrectly calling `super().npartitions`](https://github.com/dask-contrib/dask-expr/blob/4eced7384c09126800df429a6040e926477d2503/dask_expr/_expr.py#L2050) in a case where the `"npartitions"` parameter can/should not be trusted.

Perhaps we just want to change the `return super().npartitions` to `return len(self.divisions) - 1` in `PartitionsFiltered`?

Reply:
Actually, the problem is probably not in `PartitionsFiltered`, but all the way down in `Expr`! As far as I can tell, it is not really safe to assume the "npartitions" operand will always be satisfied. Therefore, we should probably simplify `Expr.npartitions`

Reply:
I don't know about that. This is a pretty good edge case and most likely only annoying when testing. I don't see this as a real world use case, I'd feel more comfortable keeping this isolated in here for now